### ISSUE TITLE
crypto feature "KVZ_CRYPTO_MVs", "KVZ_CRYPTO_MV_SIGNS"

### ISSUE THREAD

**Issue by instrol** [ultravideo/kvazaar#321]
I always get help from kvazaar members. Thank you for your hard work.
We have additional questions about crypto features "KVZ_CRYPTO_MVs" and "KVZ_CRYPTO_MV_SIGNS", so we have re-uploaded them as an issue.

As you may remember, I am trying to apply crypto features on a CU. I've made a lot of progress with the advice of the kvazaar member, but there are still unresolved issues.

![그림2](https://user-images.githubusercontent.com/28687650/160345588-ce2b8c02-e3cd-43f1-8ff6-f16be86411a1.png)

We scoped to apply the crypto features "KVZ_CRYPTO_MVs" and "KVZ_CRYPTO_MV_SIGNS" around the face. However, crypto features appear in unspecified ranges in some scenes, as in the picture. Why is this happening and is there a way to fix it?

**Comment by Jovasa** [comment]
It's not very easy to comment on inter encoded sequences based on a single frame but what I suspect is happening that the CUs outside of the face area predict pixels from the encrypted face area.

If you are able to provide an example encoded sequence I should be able to give more comments on what is happening.

**Comment by instrol** [comment]
Thanks for the reply. I actually tried to solve this problem thinking I got a hint from your answer. However, it didn't work out and I asked for additional advice.

As you said, crypto features are being applied outside of the CU range I specified.

![그림3](https://user-images.githubusercontent.com/28687650/160805756-d934240c-f7a8-4c0c-b2d2-d02f3203a3fe.png)

My attempt is to receive the coordinates of the CU from "encode_coding_tree.c" as shown in the picture and apply the crypto feature "KVZ_CRYPTO_MVs" in "cabac.c" according to the range of the CU.

In conclusion, I want the crypto feature "KVZ_CRYPTO_MVs" to be applied only within the coordinate range I specified.

**Comment by Jovasa** [comment]
![image](https://user-images.githubusercontent.com/9692471/161495043-678af764-8663-4e35-b204-159503aa044b.png)
Here is a very rough sketch of what I suspect is happening. Basically the green CU in frame 2 copies data (red area) from the previous frame which partially overlaps with the encrypted area (blue).

You need to ensure that the non encrypted CUs do not predict data from the encrypted area. You could try to change the `fracmv_within_tile` function https://github.com/ultravideo/kvazaar/blob/c755873ed237fe3d1be295a719a3feaecd12653a/src/search_inter.c#L94 to return false in these scenarios. However, there is a possibility that the encoder ends up in a situation where it cannot find a suitable movement vector and might not be able to properly deal with it.

One thing you have to pay attention is that the area where you do not want to copy data is based on the encrypted area in the **reference** frame not the current frame, which might not be trivial to do.

**Comment by instrol** [comment]
Thanks for your kind comments. Your point of the problem was perfect and when I converted all the return values of function "fracmv_within_tile" to "false", the problem was solved. The detailed implementation is up to me, so I will try my best.

Thank you very much for your help.

**Comment by Jovasa** [comment]
If you change `fracmv_within_tile` to return false always, the encoder will always use intra encoding. To verify the hypothesis you should return true for (0, 0) vector and false for everything else.

**Comment by instrol** [comment]
hello. As I reflected on your advice, an idea came to mind.

Is there a way to refer to the state of the area before encryption? The problem is caused by referencing an area that has already been encrypted. If we can refer to the state of the area before encryption, it is thought that there will be no encryption propagation.

We would appreciate it if you could tell us what you think.

thank you

**Comment by Jovasa** [comment]
The encoder already always uses the reference data before the encryption. The issue is that the decoder has no idea what the unencrypted state unless it decrypts it. You really have to have the non-encrypted CUs predict data from the non-encrypted and that is pretty much the easiest solution. The only other solution I can come up with is that you could have the re-construction buffers have always the encrypted data, but that would require modifications to both the encoder and decoder, and it would lose the main benefit of this encryption method that the bitstream is always decodable, whether it is encrypted or not.
